### DISCUSSION TITLE
How to bind Panel`s or Canvas' Children

### DISCUSSION THREAD

**Discussion by sonicsmooth** [AvaloniaUI/Avalonia#10636] [General]
I need to place some shapes at arbitrary locations pulled from a database or other collection. 

I've derived from `XYPanel` class as per https://docs.avaloniaui.net/docs/layout/create-a-custom-panel

    public class XYPanel : Panel
    {
        protected override Size MeasureOverride(Size availableSize) ...
        protected override Size ArrangeOverride(Size finalSize) ...
    }

But I don't see a way to bind the children to a collection in xaml.  
I can do this:

    <controls:XYPanel>
        <Ellipse Width="30" Height="20" Fill="Red"/>
        <Ellipse Width="20" Height="30" Fill="Blue"/>
    </controls:XYPanel>

but not this:

    <controls:XYPanel Children="{Binding MyChildren}"/>

I could populate this in code-behind, but it seems this should be doable in xaml.
Same with trying to use `Canvas` which appears designed for this purpose -- how to populate the children? 

BTW I'd pay $100 for a good and proper book about all this explaining xaml theory.

thanks

**Comment by jp2masa** [comment]
The easiest way is probably using an items control with a style, something like:

```xaml
<ItemsControl Classes="items" Items="{Binding Items"}>
    <ItemsControl.Styles>
        <Style Selector="ItemsControl.items > ContentControl">
            <Setter Property="Canvas.Top" Value="{Binding Top}" />
        </Style>
    </ItemsControl.Styles>
    <ItemsControl.ItemsPanel>
        <ItemsPanelTemplate>
            <Canvas />
        </ItemsPanelTemplate>
    </ItemsControl.ItemsPanel>
    <ItemsControl.ItemTemplate>
        <DataTemplate>
            <Ellipse Width="{Binding Width}" Height="{Binding Height}" Fill="{Binding Fill}"/>
        </DataTemplate>
    </ItemsControl.ItemTemplate>
</ItemsControl>
```

**Reply by sonicsmooth** [reply]
Thank you!  I started to go down this route but couldn't figure it out.
So instead of Ellipse, how do I bind the real items from the viewmodel?  The viewmodel may contain other shapes besides ellipse, and eventually will contain groups of such shapes.

**Reply by jp2masa** [reply]
You can for example have different viewmodels (e.g. `EllipseViewModel`, `RectangleViewModel`,...), then define `DataTemplates` for each of them, so instead of `ItemsControl.ItemTemplate` you would have something like:

```xaml
    <ItemsControl.DataTemplates>
        <DataTemplate DataType="vm:EllipseViewModel">
            <Ellipse Width="{Binding Width}" Height="{Binding Height}" Fill="{Binding Fill}"/>
        </DataTemplate>
        <DataTemplate DataType="vm:RectangleViewModel">
            <Rectangle Width="{Binding Width}" Height="{Binding Height}" Fill="{Binding Fill}"/>
        </DataTemplate>
    </ItemsControl.DataTemplates>
```
